<!DOCTYPE html>
<html lang="en-us">
  
  <head>
  <meta charset="UTF-8">
  <title>DragoNN</title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="theme-color" content="#157878">
  <link rel="stylesheet" href="/dragonn/css/normalize.css">
  <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="/dragonn/css/cayman.css">

  <!-- google analytics -->
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
      ga('create', 'UA-79329804-1', 'auto');
      ga('send', 'pageview');
  </script>

</head>


  <body>
    <section class="page-header">
  <h1 class="project-name">DragoNN</h1>
  <h2 class="project-tagline">DragoNN provides a toolkit to learn how to model and interpret regulatory sequence data using deep learning.</h2>
  <a href="/dragonn/index.html" class="btn">Overview</a>
  <a href="/dragonn/tutorials.html" class="btn">Tutorials</a>
  <a href="/dragonn/cloud_resources.html" class="btn">Cloud Resources</a>
  <a href="/dragonn/code.html" class="btn">Code</a>
  <a href="/dragonn/documentation.html" class="btn">Documentation</a>
  <a href="/dragonn/workshops.html" class="btn">Workshops</a>
  <a href="/dragonn/paper_supplement.html" class="btn">Paper Supplement</a>
</section>



    <section class="main-content">
      
      <h1 id="overview">Overview</h1>

<p>Web-based DragoNN tutorials use Jupyter notebooks to allow interactive model manipulation and visualization. The DragoNN tutorials demonstrate model design and interpretation principles using simulated data. Having access to the ground truth in the simulations enables us to dissect the specific parameters affecting prediction performance and model interpretation. Simulations can thus enable systematic evaluation of model design and interpretation to quickly find principles that could effectively apply to real genomic data.
<img src="/dragonn/images/dragonn_workflow.png" alt="dragonn_workflow" /></p>

<h1 id="table-of-contents">Table of Contents</h1>
<ul>
  <li><a href="#getting-access-to-dragonn-software-and-gpu-hardware]">Configuring software and hardware</a></li>
  <li><a href="#online-dragonn-course">Online DragoNN course ()</a></li>
  <li><a href="#intro-tutorial">Intro tutorial (GPU not required)</a></li>
  <li><a href="#intermediate-tutorial">Intermediate Tutorial (GPU recommended)</a></li>
</ul>

<h1 id="getting-access-to-dragonn-software-and-gpu-hardware">Getting access to DragoNN software and GPU hardware</h1>

<p>Each tutorial below specifies software and hardware requirements.</p>

<p>Tutorials with software requirements will necessitate either a released version of the DragoNN package or the bleeding-edge (non-released) version. See the <a href="/dragonn/code.html">code page</a> for software installation and configuration instructions.</p>

<p>Tutorials with hardware requirements will necessitate either a CPU (Tutorial 1) or a GPU (Tutorials 2 - 5). If the tutorial requires a GPU and you do not have access to GPUs locally, you can run the tutorials in Google Colaboratory (see below) with a GPU accelerator. 

<h1 id="google-colab">Google Colaboratory Tutorials</h1>
<p>The following notebooks can be opened and executed with <a href="https://colab.research.google.com/notebooks/welcome.ipynb#recent=true"/>Google Colaboratory</a></p>
<p><a href="https://colab.research.google.com/github/kundajelab/dragonn/blob/master/tutorials/PrimerTutorial%201%20-%20Exploring%20model%20architectures%20for%20a%20homotypic%20motif%20density%20simulation.ipynb">Tutorial 1 from Supplement of DragoNN Primer: Exploring convolutional neural network (CNN) architectures for simulated genomic data.
</a></p>
<p><a href="https://colab.research.google.com/github/kundajelab/dragonn/blob/master/tutorials/PrimerTutorial%202%20-%20Hyperparameter%20Tuning%20via%20Grid%20Search.ipynb">Tutorial 2 from Supplement of DragoNN Primer: CNN hyperparameter tuning via grid search.
</a></p>
<p><a href="https://colab.research.google.com/github/kundajelab/dragonn/blob/master/tutorials/PrimerTutorial%203%20-%20Interpreting%20features%20induced%20by%20DNN's%20across%20multiple%20types%20of%20motif%20grammars.ipynb">Tutorial 3 from Supplement of DragoNN Primer:Interpreting features induced by CNN's across multiple types of motif grammars.</a></p> 

<p><a href="https://colab.research.google.com/github/kundajelab/dragonn/blob/master/tutorials/PrimerTutorial%204%20-%20Interpreting%20predictive%20sequence%20features%20in%20%20TF%20binding%20events%20within%20the%20GM12878%20cell%20line.ipynb">Tutorial 4 from Supplement of DragoNN Primer:  Interpreting predictive sequence features in  TF binding events within the GM12878 cell line. </a></p>

<p><a href="https://colab.research.google.com/github/kundajelab/dragonn/blob/master/tutorials/PrimerTutorial%205%20-%20Functional%20variant%20characterization%20for%20non-coding%20SNPs%20within%20the%20SPI1%20motif.ipynb">Tutorial 5 from Supplement of Dragonn Primer: Functional variant characterization for non-coding SNPs within the SPI1 motif.</a></p>

<h1 id="intro-tutorial">Intro tutorials included with the dragonn source code</h1>
<p><em>Software requirements: bleeding-edge DragoNN package.</em></p>

<p><em>Hardware requirements: CPU/laptop sufficient for tutorial 1. GPU hardware required for tutorials 2 - 5.</em></p>

<p>To explore the Jupyter notebook tutorials, navigate to the <code class="highlighter-rouge">./tutorials</code> directory and run</p>

<div class="highlighter-rouge"><pre class="highlight"><code>jupyter notebook 
</code></pre>
</div>

<p>This will start a jupter notebook server, allowing you to navigate to the tutorial notebooks in your browser:
  <ul>
    <li><a href="https://github.com/kundajelab/dragonn/blob/master/paper_supplement/PrimerTutorial%201%20-%20Exploring%20model%20architectures%20for%20a%20homotypic%20motif%20density%20simulation.ipynb">Tutorial 1: Exploring convolutional neural network (CNN) architectures for simulated genomic data</a></li>
    <li><a href="https://github.com/kundajelab/dragonn/blob/master/paper_supplement/PrimerTutorial%202%20-%20Hyperparameter%20Tuning%20via%20Grid%20Search.ipynb">CNN Hyperparameter Tuning via Grid Search</a></li>
    <li><a href="https://github.com/kundajelab/dragonn/blob/master/paper_supplement/PrimerTutorial%203%20-%20Interpreting%20features%20induced%20by%20DNN's%20across%20multiple%20types%20of%20motif%20grammars.ipynb">Interpreting features induced by DNN's across multiple types of motif grammars</a></li>
    <li><a href="https://github.com/kundajelab/dragonn/blob/master/paper_supplement/PrimerTutorial%204%20-%20Interpreting%20predictive%20sequence%20features%20in%20%20TF%20binding%20events%20within%20the%20GM12878%20cell%20line.ipynb">Interpreting predictive sequence features in TF binding events within the GM12878 cell line.</a></li>
    <li><a href="https://github.com/kundajelab/dragonn/blob/master/paper_supplement/PrimerTutorial%205%20-%20Functional%20variant%20characterization%20for%20non-coding%20SNPs%20within%20the%20SPI1%20motif.ipynb">Functional variant characterization for non-coding SNPs within the SPI1 motif</a></li>


      <footer class="site-footer">
  <span class="site-footer-owner"><a href="">DragoNN</a> is maintained by <a href="https://www.linkedin.com/in/jisraeli/">Johnny Israeli</a> in the  <a href="https://github.com/kundajelab">Kundaje Lab</a>.</span>
  <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/pietromenna/jekyll-cayman-theme">Cayman theme</a> by <a href="http://github.com/jasonlong">Jason Long</a>.</span>
</footer>


    </section>

  </body>
</html>
